README.md

```diff
@@ -1 +1,467 @@
+
+## array (list, sequence, vector) <= pick one
+
+À vue de nez il nous faudrait au minimum un autre type : le tableau pour pouvoir faire une chaîne de nombre.
+Si ensuite on fait les strings comme des suites de caractères, et les caractères sont des nombres (des octets), le tableau pourrait être réutilisé pour implémenter les strings.
+Les tableaux pourraient aussi servir aux I/O de fonctions (arguments en entrée, valeurs de retours en sortie).
+
+### bonus
+
+Autre questionnements :
+
+- un tableau est indexé à 0 par défaut, est-ce qu'on ne permettrait pas de changer ça ?
+- puisque qu'un tableau est par définition associatif avec des entiers naturels en tant que clefs, est-ce qu'on ne mixerait pas ce concept (associative array, map, dictionnary, hashmap, key-value store, c'est la même chose essentiellement) avec le tableau ?
+
+Il y a des similitudes entre un tableau classique et une map :
+
+```
+a = [2, 4, 8, 16]
+b = {0:2, 1:4, 2:8, 3:16}
+a[0] == b[0] && a[3] == b[3] // true
+```
+
+Lua l'a fait avec la notion de table. À voir ce que ça implique mais cette idée de généraliser le concept me tente bien.
+
+## pointeur
+
+Histoire de laisser le dev choisir s'il fait une copie ou une ref ? Si on l'implémente est-ce qu'on n'interdit pas `null` ? Je ne vois pas l'utilité de `None` / `undefined` / `null`, etc
+
+## fonction
+
+Indispensable de pouvoir gérer une fonction en tant que valeur n'est-ce pas ?
+
+## symbol / enumeration
+
+Dans la plupart des langages cela se fait avec des constantes (comme en Go), mais permet de mélanger les concepts (c'est au dev de faire gaffe ?) :
+
+```
+color [white, black] // 0, 1
+theme [dark, light] // 0, 1
+
+Colorize(dark) // valid but wrong
+```
+
+Je propose de ne pas créer de nouveau type, un enum c'est en fait un tableau de labels/identifiants (indexé par défaut à 0, mais on peut réindexer chaque élément)
+
+### liens
+
+https://en.wikipedia.org/wiki/Enumerated_type
+
+## error
+
+Est-ce qu'on ne dirait pas que c'est un label qui doit être défini qu'on throw/raise ? Avec possibilité d'ajouter de la donnée (comme ça tout est permis).
+
+```
+// ceci est un enum, donc juste un tableau d'identifiants
+IO_errors [disk_full, network_unreachable, write_error]
+
+[...]
+
+  // error message
+  raise write_error, `not written: ` & written
+
+  // just give the data and let the caller format their own error message
+  raise write_error, to_write - written
+```
+
+Ensuite le caller peut faire le tri dans l'exception qu'il choppe (en plus bien sûr de la tester directement) :
+
+```
+if error in IO_errors then
+  ...
+```
```

Comment:
Jpense que d'utiliser les `effect` pour ca me semble une bonne idee.

Ca nous premet d'avoir un seul concept pour qui permet d'exprimer pleins de features plutard sans avoir a rajouter de syntaxe specifique.

J'me renseigne sur le sujet j'suis entrain de lire https://drops.dagstuhl.de/opus/volltexte/2018/9762/pdf/dagrep_v008_i004_p104_18172.pdf

> Algebraic effect handlers generalise many control-flow abstractions such as exception handling, iterators, `async/await`, or backtracking, and in turn allow them to be expressed as libraries rather than implementing them as primitives as many language implementations do

Reply:
j'ai lu le truc c'est de la merde c'est juste des ennoncer de sujets qu'ils on traiter pendant leur workshop mais ya pas de contenu, j'vais continuer a explorer un peu le concept voir si ca apporte un truc vraiment ou si c'est ce prendre la tete pour rien.